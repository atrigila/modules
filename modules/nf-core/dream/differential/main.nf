process DREAM_DIFFERENTIAL {
    tag "${meta.id} - ${meta.contrast_id}"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/d6/d6fa8a7908dd357484d63bb574a378f1739440a2c0752ce91b1e0e9d1ac1638c/data' :
        'community.wave.seqera.io/library/bioconductor-edger_bioconductor-variancepartition_r-optparse:ba778938d72f30c5' }"

    input:
    tuple val(meta), path(samplesheet), path(counts)

    output:
    tuple val(meta), path("*.dream.results.tsv")        , emit: results
    tuple val(meta), path("*.dream.mean_difference.png"), emit: md_plot
    tuple val(meta), path("*.MArrayMM.dream.rds")       , emit: rdata
    tuple val(meta), path("*.dream.model.txt")          , emit: model
    tuple val(meta), path('*.dream.contrasts_plot.png') , emit: contrasts_png
    tuple val(meta), path("*.R_sessionInfo.log")        , emit: session_info
    tuple val(meta), path("*.normalised_counts.tsv")    , emit: normalised_counts, optional: true
    path "versions.yml"                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    template 'dream_de.R'

    stub:
    prefix = task.ext.prefix   ?: "${meta.id}"
    """
    touch "${meta.contrast_id}.dream.results.tsv"
    touch "${meta.contrast_id}.dream.mean_difference.png"
    touch "${meta.contrast_id}.dream.contrasts_plot.png"
    touch "${meta.contrast_id}.MArrayMM.dream.rds"
    touch "${meta.contrast_id}.dream.model.txt"
    touch "${meta.contrast_id}.R_sessionInfo.log"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        r-edger: \$(Rscript -e "library(edgeR); cat(as.character(packageVersion('edgeR')))")
        r-variancepartition: \$(Rscript -e "library(variancePartition); cat(as.character(packageVersion('variancePartition')))")
    END_VERSIONS
    """
}
