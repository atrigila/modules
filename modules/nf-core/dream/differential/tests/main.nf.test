nextflow_process {

    name "Test Process DREAM_DIFFERENTIAL"
    script "../main.nf"
    process "DREAM_DIFFERENTIAL"

    tag "modules"
    tag "models"
    tag "dge"
    tag "dream"
    tag "dream/differential"

    test("variancePartition data - contrasts + blocking factor") {

        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [
                            contrast_id: 'Disease_0_1',
                            contrast_variable: 'Disease',
                            contrast_reference: '0',
                            contrast_target: '1',
                            blocking_factors: 'Individual'
                        ],
                        file("https://raw.githubusercontent.com/atrigila/test-datasets/refs/heads/modules/data/delete_me/variancepartition/metadata.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/atrigila/test-datasets/refs/heads/modules/data/delete_me/variancepartition/countmatrix.csv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Mus musculus - expression table - contrasts") {

        when {
            process {
                """
                // contrast data + counts + samplesheet
                input[0] = Channel.of(
                    [
                        [
                            contrast_id: 'treatment_mCherry_hND6',
                            contrast_variable: 'treatment',
                            contrast_reference: 'mCherry',
                            contrast_target: 'hND6',
                            blocking_factors: null
                        ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Mus musculus - expression table - contrasts + blocking factors") {

        when {
            process {
                """
                // contrast data + counts + samplesheet
                input[0] = Channel.of(
                    [
                        [
                            contrast_id: 'treatment_mCherry_hND6',
                            contrast_variable: 'treatment',
                            contrast_reference: 'mCherry',
                            contrast_target: 'hND6',
                            blocking_factors: 'sample_number'
                        ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Mus musculus - expression table - contrasts + blocking factors stub") {

        options '-stub'
        when {
            process {
                """
                // contrast data + counts + samplesheet
                input[0] = Channel.of(
                    [
                        [
                            contrast_id: 'treatment_mCherry_hND6',
                            contrast_variable: 'treatment',
                            contrast_reference: 'mCherry',
                            contrast_target: 'hND6',
                            blocking_factors: 'sample_number'
                        ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv", checkIfExists: true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/refs/heads/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}