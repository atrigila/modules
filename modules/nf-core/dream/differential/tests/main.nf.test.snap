{
    "Mus musculus - expression table - contrasts + blocking factors": {
        "content": [
            {
                "0": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,15129ab903059d9c3e3f54d7ae5ae3d3"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a0c919600304e2aa1a474374f897cc8b"
                ],
                "results": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,15129ab903059d9c3e3f54d7ae5ae3d3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a0c919600304e2aa1a474374f897cc8b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-10T17:34:19.160419833"
    },
    "Mus musculus - expression table - contrasts + blocking factors stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,ad2f8b01aadf7f55eb6341db91322ab0"
                ],
                "results": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": "sample_number"
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ad2f8b01aadf7f55eb6341db91322ab0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-10T17:34:43.471218203"
    },
    "Mus musculus - expression table - contrasts": {
        "content": [
            {
                "0": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,6c172497a3ade420037e5abcb29cae2e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,a0c919600304e2aa1a474374f897cc8b"
                ],
                "results": [
                    [
                        {
                            "contrast_id": "treatment_mCherry_hND6",
                            "contrast_variable": "treatment",
                            "contrast_reference": "mCherry",
                            "contrast_target": "hND6",
                            "blocking_factors": null
                        },
                        "treatment_mCherry_hND6.dream.results.tsv:md5,6c172497a3ade420037e5abcb29cae2e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a0c919600304e2aa1a474374f897cc8b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-10T17:30:59.943799883"
    }
}